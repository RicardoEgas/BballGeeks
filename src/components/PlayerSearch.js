import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Autosuggest from 'react-autosuggest';
import {
  fetchPlayers,
  selectPlayers,
  selectPlayersStatus,
  selectPlayersError,
} from '../redux/players';
import PlayerStats from './PlayerStats';

function PlayerSearch() {
  const dispatch = useDispatch();
  const status = useSelector(selectPlayersStatus);
  const players = useSelector(selectPlayers);
  const error = useSelector(selectPlayersError);

  const [searchTerm, setSearchTerm] = useState('');
  const [suggestions, setSuggestions] = useState([]);
  const [selectedPlayer, setSelectedPlayer] = useState(null);

  useEffect(() => {
    // Fetch player data using Redux thunk
    dispatch(fetchPlayers());
  }, [dispatch]);

  const getSuggestions = (value) => {
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;

    return inputLength === 0
      ? []
      : players.filter(
        (player) => player.first_name.toLowerCase().includes(inputValue)
            || player.last_name.toLowerCase().includes(inputValue),
      );
  };

  const onSuggestionsFetchRequested = ({ value }) => {
    setSuggestions(getSuggestions(value));
  };

  const onSuggestionsClearRequested = () => {
    setSuggestions([]);
  };

  const getSuggestionValue = (suggestion) => `${suggestion.first_name} ${suggestion.last_name}`;

  const renderSuggestion = (suggestion) => (
    <div className="suggestions-list-item">{`${suggestion.first_name} ${suggestion.last_name}`}</div>
  );

  const handleSuggestionSelected = (_, { suggestionValue }) => {
    const selected = players.find(
      (player) => `${player.first_name} ${player.last_name}` === suggestionValue,
    );
    setSelectedPlayer(selected.id);
  };

  if (status !== 'succeeded') {
    return (
      <div>
        <h1>Player Search</h1>
        {status === 'loading' && (
          <p>Loading data, please wait a few seconds...</p>
        )}
        {status === 'failed' && (
          <p>
            Error:
            {error}
          </p>
        )}
      </div>
    );
  }

  return (
    <div>
      <h1>Player Search</h1>
      <p>You can search a player now! ðŸ˜ƒ </p>
      <Autosuggest
        suggestions={suggestions}
        onSuggestionsFetchRequested={onSuggestionsFetchRequested}
        onSuggestionsClearRequested={onSuggestionsClearRequested}
        getSuggestionValue={getSuggestionValue}
        renderSuggestion={renderSuggestion}
        inputProps={{
          className: 'autosuggest-input',
          placeholder: 'Search for a player',
          value: searchTerm,
          onChange: (_, { newValue }) => setSearchTerm(newValue),
        }}
        onSuggestionSelected={handleSuggestionSelected}
      />
      {selectedPlayer && <PlayerStats selectedPlayerId={selectedPlayer} />}
    </div>
  );
}

export default PlayerSearch;
