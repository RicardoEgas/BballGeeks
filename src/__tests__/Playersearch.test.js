import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import PlayerSearch from '../components/PlayerSearch';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('PlayerSearch component', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      players: {
        status: 'succeeded',
        error: null,
        data: [],
      },
    });
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <PlayerSearch />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
