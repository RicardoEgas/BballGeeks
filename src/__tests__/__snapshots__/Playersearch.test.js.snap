// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerSearch component renders correctly 1`] = `
<div>
  <h1>
    Player Search
  </h1>
  <p>
    You can search a player now! ðŸ˜ƒ 
  </p>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-owns="react-autowhatever-1"
    className="react-autosuggest__container"
    role="combobox"
  >
    <input
      aria-activedescendant={null}
      aria-autocomplete="list"
      aria-controls="react-autowhatever-1"
      autoComplete="off"
      className="autosuggest-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="Search for a player"
      type="text"
      value=""
    />
    <div
      className="react-autosuggest__suggestions-container"
      id="react-autowhatever-1"
      role="listbox"
    />
  </div>
</div>
`;
